File: minify.js
// Simple minifier script (runs with Node.js)
// Usage: node minify.js

const fs = require('fs');
const path = require('path');

// Read the original script
const scriptPath = path.join(__dirname, 'web', 'script.js');
const scriptContent = fs.readFileSync(scriptPath, 'utf8');

// Simple minification (remove comments, extra whitespace, etc.)
function simpleMinify(code) {
    // Remove single-line comments (but keep URLs and important comments)
    code = code.replace(/\/\/.*$/gm, '');
    
    // Remove multi-line comments (but be careful with regex)
    code = code.replace(/\/\*[\s\S]*?\*\//g, '');
    
    // Remove extra whitespace
    code = code.replace(/\s+/g, ' ');
    
    // Remove whitespace around operators and brackets
    code = code.replace(/\s*([{}();,=+\-*/%<>!&|?:])\s*/g, '$1');
    
    // Remove whitespace at start/end of lines
    code = code.replace(/^\s+|\s+$/gm, '');
    
    // Remove multiple consecutive spaces
    code = code.replace(/\s{2,}/g, ' ');
    
    // Remove whitespace before semicolons
    code = code.replace(/\s*;/g, ';');
    
    // Remove whitespace after opening brackets and before closing brackets
    code = code.replace(/\s*{\s*/g, '{');
    code = code.replace(/\s*}\s*/g, '}');
    code = code.replace(/\s*\(\s*/g, '(');
    code = code.replace(/\s*\)\s*/g, ')');
    
    return code.trim();
}

// Obfuscate variable names (simple approach - just shorten common patterns)
function obfuscate(code) {
    // This is a very basic obfuscation - for better results use a proper tool
    // But it makes the code harder to read
    
    // Replace common patterns with shorter names
    const replacements = {
        'gameState': 'g',
        'this.gameState': 'this.g',
        'allParties': 'p',
        'budgetSubjects': 'b',
        'currentDay': 'd',
        'partyControlPercentage': 'pc',
        'socialValue': 'sv',
        'economicValue': 'ev'
    };
    
    // Only replace in safe contexts (not in strings)
    let obfuscated = code;
    for (const [original, replacement] of Object.entries(replacements)) {
        // Replace whole word matches only
        const regex = new RegExp(`\\b${original}\\b`, 'g');
        obfuscated = obfuscated.replace(regex, replacement);
    }
    
    return obfuscated;
}

// Minify the code
console.log('Minifying script.js...');
const minified = simpleMinify(scriptContent);

// Optional: Add basic obfuscation
console.log('Adding basic obfuscation...');
const obfuscated = obfuscate(minified);

// Write minified version
const outputPath = path.join(__dirname, 'web', 'script.min.js');
fs.writeFileSync(outputPath, obfuscated, 'utf8');

console.log(`✓ Minified script written to: ${outputPath}`);
console.log(`  Original size: ${scriptContent.length} bytes`);
console.log(`  Minified size: ${obfuscated.length} bytes`);
console.log(`  Reduction: ${((1 - obfuscated.length / scriptContent.length) * 100).toFixed(1)}%`);

File: minify.py
#!/usr/bin/env python3
"""
Simple JavaScript minifier
Usage: python minify.py
"""

import re
import os

def minify_js(code):
    """Minify JavaScript code by removing comments and whitespace"""
    
    # Remove single-line comments (but preserve URLs)
    code = re.sub(r'//[^\n]*', '', code)
    
    # Remove multi-line comments
    code = re.sub(r'/\*.*?\*/', '', code, flags=re.DOTALL)
    
    # Remove extra whitespace (but keep spaces in strings)
    # First, protect strings
    strings = []
    def protect_string(match):
        strings.append(match.group(0))
        return f'__STRING_{len(strings)-1}__'
    
    code = re.sub(r'["\'][^"\']*["\']', protect_string, code)
    
    # Remove whitespace around operators and brackets
    code = re.sub(r'\s+', ' ', code)
    code = re.sub(r'\s*([{}();,=+\-*/%<>!&|?:\[\]])\s*', r'\1', code)
    code = re.sub(r'\s+', ' ', code)
    
    # Restore strings
    for i, string in enumerate(strings):
        code = code.replace(f'__STRING_{i}__', string)
    
    return code.strip()

def main():
    script_path = os.path.join('web', 'script.js')
    
    if not os.path.exists(script_path):
        print(f"Error: {script_path} not found!")
        return
    
    print(f"Reading {script_path}...")
    with open(script_path, 'r', encoding='utf-8') as f:
        original = f.read()
    
    print("Minifying...")
    minified = minify_js(original)
    
    output_path = os.path.join('web', 'script.min.js')
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(minified)
    
    original_size = len(original)
    minified_size = len(minified)
    reduction = (1 - minified_size / original_size) * 100
    
    print(f"\n[OK] Minified script written to: {output_path}")
    print(f"  Original size: {original_size:,} bytes")
    print(f"  Minified size: {minified_size:,} bytes")
    print(f"  Reduction: {reduction:.1f}%")
    print(f"\n[!] Don't forget to update index.html to use script.min.js!")

if __name__ == '__main__':
    main()

